### PR TITLE
Add setPreRender/setPostRender hooks & example

### PR DESCRIPTION
**Pull request by shilman**
## What I did

- [x] Added test lifecycle hooks
- [x] Added image snapshotting example & snapshots
- [x] Added initial documentation
- [x] Aligned stories.json/CSF mode behavior

```js
setPostRender(async (page, context) => {
  const image = await page.screenshot();
  expect(image).toMatchImageSnapshot({
    customSnapshotsDir,
    customSnapshotIdentifier: context.id,
  });
});
```

## How to test

```
yarn build
yarn test-storybook:playwright
# edit a component story
yarn test-storybook:playwright
```

<details>
  <summary>📦 Published PR as canary version: <code>0.0.2-canary.38.ad367be.0</code></summary>
  <br />
  
  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @storybook/test-runner@0.0.2-canary.38.ad367be.0
  # or 
  yarn add @storybook/test-runner@0.0.2-canary.38.ad367be.0
  ```
</details>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -228,6 +229,55 @@ jobs:
...
+```js
+it('component--widget', async () => {
+  const page = newPage();
+  const context = { id: 'component--widget', title: 'Component', name: 'Widget' };
+  await page.goto(STORYBOOK_URL);
+
+  // pre-render hook
+  if (preRender) await preRender(page, context);
+
+  // render the story and run its paly function (if applicable)
```
Comment (yannbf) [inline README.md:249]:
```suggestion
  // render the story and run its play function (if applicable)
```

File: README.md [hunk 2]
```diff
@@ -228,6 +229,55 @@ jobs:
...
+  await page.execute('render', context);
+
+  // post-render hook
+  if (postRender) await postRender(page, context);
+});
+```
+
+The hooks here, `preRender` and `postRender` are functions that take a [Playwright Page](https://playwright.dev/docs/pages) and a context object with the current story `id`, `title`, and `name`. They are globally settable by `@storybook/test-runner`'s `setPreRender` and `setPostRender` APIs.
+
+Thus, to make the test runner perform image snapshotting, you might set up the following in your `jest-setup.js`:
```
Comment (yannbf) [inline README.md:259]: I think we need to add a better description here.

People need to:
1 - "Eject" the config file
2 - Add a [setupFileAfterEnv](https://jestjs.io/docs/configuration#setupfilesafterenv-array) section
3 - Add the jest-setup.js (or whatever other name) file to the project.

Would it make sense to have some sort of "eject" command? So we generate the file override for the users instead?

I'm fine with not doing any of that and just adding a link to the jest setup file docs, but the average user might get lost in the instructions

File: src/csf/transformCsf.ts [hunk 3]
```diff
@@ -39,7 +40,7 @@ const prefixFunction = (
   };
 
   // instead, let's just make the prefixer return the function
```
Comment (yannbf) [inline src/csf/transformCsf.ts:42]: does this comment still make sense?

File: README.md (hunk 2 above)
Comment (shilman) [reply README.md:259]: Eject command sounds great!

Comment (shilman) [reply README.md:259]: Can you handle it @yannbf ? https://github.com/storybookjs/test-runner/issues/42

Comment (yannbf) [reply README.md:259]: Sure!

File: README.md [hunk 4]
```diff
@@ -228,6 +229,55 @@ jobs:
...
+
+The test runner renders a story and executes its [play function](https://storybook.js.org/docs/react/writing-stories/play-function) if one exists. However, there are certain behaviors that are not possible to achieve via the play function, which executes in the browser. For example, if you want the test runner to take visual snapshots for you, this is something that is possible via Playwright/Jest, but must be executed in Node.
+
+To enable use cases like visual or DOM snapshots, the test runner exports test hooks that can be overridden globally. These hooks give you access to the test lifecycle before and after the story is rendered.
+
+Consider the following pseudocode:
+
+```js
+it('component--widget', async () => {
+  const page = newPage();
```
Comment (kylegach) [inline README.md:242]: What is `newPage`? Should it be `new Page()`?

File: README.md [hunk 5]
```diff
@@ -228,6 +229,55 @@ jobs:
...
+## Experimental test hook API
+
+The test runner renders a story and executes its [play function](https://storybook.js.org/docs/react/writing-stories/play-function) if one exists. However, there are certain behaviors that are not possible to achieve via the play function, which executes in the browser. For example, if you want the test runner to take visual snapshots for you, this is something that is possible via Playwright/Jest, but must be executed in Node.
+
+To enable use cases like visual or DOM snapshots, the test runner exports test hooks that can be overridden globally. These hooks give you access to the test lifecycle before and after the story is rendered.
+
+Consider the following pseudocode:
+
+```js
+it('component--widget', async () => {
```
Comment (kylegach) [inline README.md:241]: Is this pseudo-code, written per-test, necessary _along with_ the more global code in the next snippet?

File: README.md (hunk 4 above)
Comment (shilman) [reply README.md:242]: Pseudo code

File: README.md (hunk 5 above)
Comment (shilman) [reply README.md:241]: Thanks I'll rewrite this part to clarify

**Comment by shilman** [comment]
@p01 Thanks for reaching out and very nice to meet you! Storywright looks really cool--it's the first time I'm seeing it. I'd love to learn more about where you're going with it, talk through what we're aiming for with the test runner, and figure out how we might collaborate. Any interest in doing a call in the next couple weeks?

File: README.md (hunk 5 above)
Comment (shilman) [reply README.md]: @kylegach updated to (hopefully) make it more clear

Comment (kylegach) [reply README.md]: Yes, that makes it _much_ more clear! I wasn't grokking the "automatically generated" nature of it, before.

**Pull request merged**